src/BSON/Int64.c

```diff
@@ -244,6 +263,100 @@ static int php_phongo_int64_compare_objects(zval* o1, zval* o2)
 	return 0;
 }
 
+static zend_result php_phongo_int64_cast_object(phongo_compat_object_handler_type* readobj, zval* retval, int type)
+{
+	php_phongo_int64_t* intern;
+
+	intern = Z_OBJ_INT64(PHONGO_COMPAT_GET_OBJ(readobj));
+
+	switch (type) {
+		case IS_LONG:
+#if PHP_VERSION_ID >= 70300
+		case _IS_NUMBER:
+#endif
+#if SIZEOF_ZEND_LONG == 4
+			if (intern->integer > INT32_MAX || intern->integer < INT32_MIN) {
+				zend_error(E_WARNING, "Truncating 64-bit integer value %" PRId64 " to 32 bits", intern->integer);
+			}
+#endif
+
+			ZVAL_LONG(retval, intern->integer);
+
+			return SUCCESS;
+
+		case _IS_BOOL:
+			ZVAL_BOOL(retval, intern->integer != 0);
+
+			return SUCCESS;
+
+		default:
+			return zend_std_cast_object_tostring(readobj, retval, type);
+	}
+}
+
+static zend_result php_phongo_int64_do_operation(zend_uchar opcode, zval* result, zval* op1, zval* op2)
+{
+	int64_t value1, value2;
+
+	if (Z_TYPE_P(op1) == IS_LONG) {
+		value1 = Z_LVAL_P(op1);
+	} else if (Z_TYPE_P(op1) == IS_OBJECT && Z_OBJCE_P(op1) == php_phongo_int64_ce) {
+		value1 = Z_INT64_OBJ_P(op1)->integer;
+	} else {
+		return FAILURE;
+	}
+
+	if (Z_TYPE_P(op2) == IS_LONG) {
+		value2 = Z_LVAL_P(op2);
+	} else if (Z_TYPE_P(op2) == IS_OBJECT && Z_OBJCE_P(op2) == php_phongo_int64_ce) {
+		value2 = Z_INT64_OBJ_P(op2)->integer;
+	} else {
+		return FAILURE;
+	}
+
+	switch (opcode) {
+		case ZEND_ADD:
+			ZVAL_INT64_OBJ(result, value1 + value2);
```

Comment:
This has taken me a lot longer than I'd like to admit, but I finally got somewhere. Here's what's going on:
* For `ZEND_ADD` and `ZEND_SUB` I extracted the method used in [`fast_long_add_function`](https://github.com/php/php-src/blob/7304b56f11771ba4ed51c0053c47b94c1b9bb9a3/Zend/zend_operators.h#L699..L700) to check for a potential overflow and return the result as a double.
* For `ZEND_MUL`, I duplicated the behaviour used in [`ZEND_SIGNED_MULTIPLY_LONG`](https://github.com/php/php-src/blob/master/Zend/zend_multiply.h#L112..L115) which compares the int64 result to the result of the same operation in a `long double`, which has enough precision to handle the int64 range.
* For `ZEND_DIV`, the only way it can overflow is the case of `INT64_MIN / -1`. I explicitly covered that case and also found an issue where the result was always treated as an int even if it should've been a float. This is now properly covered by tests.
* `ZEND_SL` does not produce overflows ([see 3v4l.org](https://3v4l.org/uksb3)), and I added a check that `INT64_MAX << 1` does indeed not overflow
* `ZEND_SR` cannot overflow or underflow by definition as it shifts everything to the right by 1 bit
* `ZEND_POW` was fixed accordingly.